### PR TITLE
feat: seed explicit LS change tracker from pre-existing global settings [IDE-2067]

### PR DESCRIPTION
**Pull request by bastiandoetsch**
## Summary

- Add `seedExplicitChangesFromExistingSettings()` to `explicitLsKeyTracking.ts` — runs once after tracker construction, inspects each `SETTINGS_REGISTRY` entry, and marks any key whose VS Code global value differs from its registered default as explicitly changed
- Fixes: on fresh install/upgrade `ExplicitLspConfigurationChangeTracker` is empty, so all non-`alwaysChanged` settings go out with `changed:false` on first `workspace/didChangeConfiguration`, causing the LS to silently ignore pre-existing user-customised values
- Uses lodash `isEqual` for deep equality; guards `globalValue === undefined || defaultValue === undefined` to avoid false positives; idempotent across activations
- Mirrors the fix applied to the Eclipse plugin (PR #395)

## Test plan

- [ ] `npm run test:unit` — 273 passing (10 new: T1–T9 + T_F2 covering all guard rules)
- [ ] Manual T10: fresh-install with customised global endpoint → first `didChangeConfiguration` honoured by LS
- [ ] CI green

___

### **PR Type**
Enhancement, Bug fix

___

### **Description**
- Seed tracker with pre-existing user settings.

- Fixes lost customizations on fresh installs.

- Handles various configuration scenarios and edge cases.

- Includes comprehensive unit tests for seeding logic.

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A[Extension Initialization] --> B(Create Tracker);
  B --> C{Seed Tracker with User Settings};
  C --> D[LS Processes Config];
  C -- Handles: alwaysChanged, no default, diffs --> C;
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>explicitLsKeyTracking.ts</strong><dd><code>Add function to seed explicit LS change tracker</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/snyk/common/languageServer/explicitLsKeyTracking.ts

<ul><li>Introduces <code>seedExplicitChangesFromExistingSettings</code> function.<br> <li> Iterates through <code>SETTINGS_REGISTRY</code> to inspect VS Code global settings.<br> <li> Seeds the <code>ExplicitLspConfigurationChangeTracker</code> if global settings <br>differ from defaults.<br> <li> Handles edge cases like <code>alwaysChanged</code> entries and settings without a <br>default value.</ul>

</details>

  </td>
  <td><a href="https://github.com/snyk/vscode-extension/pull/763/changes#diff-4cbfc77e3e72d54d8d09f0fbb372b573fc07b00acf1ee2320ea3dcf21ed0b917">+44/-1</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>extension.ts</strong><dd><code>Integrate seeding logic into extension initialization</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/snyk/extension.ts

<ul><li>Imports the new <code>seedExplicitChangesFromExistingSettings</code> function.<br> <li> Calls <code>seedExplicitChangesFromExistingSettings</code> after initializing the <br><code>ExplicitLspConfigurationChangeTracker</code>.</ul>

</details>

  </td>
  <td><a href="https://github.com/snyk/vscode-extension/pull/763/changes#diff-cd593458f6d172e17fb40ceb7782e58f1ab4c804e40abcd54a2b3f236a020722">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>explicitLsKeyTracking.test.ts</strong><dd><code>Add unit tests for explicit LS change tracker seeding</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/test/unit/common/languageServer/explicitLsKeyTracking.test.ts

<ul><li>Adds a comprehensive unit test suite for <br><code>seedExplicitChangesFromExistingSettings</code>.<br> <li> Includes tests for various scenarios: differing global values, equal <br>values, undefined values, <code>alwaysChanged</code> entries, and no <code>vscodeKey</code> <br>entries.<br> <li> Tests idempotency and deep equality checks.<br> <li> Includes integration-style tests to verify seeded changes reflect in <br><code>LanguageServerSettings.fromConfiguration</code>.</ul>

</details>

  </td>
  <td><a href="https://github.com/snyk/vscode-extension/pull/763/changes#diff-ba06d306c9345438406299aad8e7e6feec4eee814b1764fcce4939994c26f669">+339/-0</a>&nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

**Comment by bastiandoetsch** [comment]
**Re: Missing Workspace Scope Seeding**

Workspace/WorkspaceFolder-scoped settings don't go through the `changed` flag mechanism — they travel via the `folderConfigs` field in the LSP payload (`LspConfigurationParam.folderConfigs`), which is populated by `FolderConfig.toLspFolderConfiguration()` unconditionally on every `didChangeConfiguration`. The `ExplicitLspConfigurationChangeTracker` and the `changed` flag only apply to the global settings map.

So for workspace-scoped overrides:
- If a setting has a workspace override, VS Code's `inspectConfiguration` returns a `workspaceValue` that `entry.resolve(configuration)` already incorporates — the *effective* value sent is correct regardless of the `changed` flag
- Settings that are specifically workspace/folder-level (org, scan config) are sent via `folderConfigs`, not the global settings map at all

This PR is intentionally scoped to global (User-scope) settings, which is the gap confirmed in IDE-2067 and the level at which the Eclipse fix (PR #395) operates. No additional seeding is needed.

File: src/snyk/common/languageServer/explicitLsKeyTracking.ts [hunk 1]
```diff
@@ -22,6 +25,44 @@ export function markExplicitLsKeysFromConfigurationChangeEvent(
...
+    if (!entry.vscodeKey) continue;
+    // R5: idempotent — already tracked by this or a previous activation
+    if (tracker.isExplicitlyChanged(lsKey)) continue;
+
+    const { configurationId, section } = Configuration.getConfigName(entry.vscodeKey);
+    const inspect = workspace.inspectConfiguration(configurationId, section);
+
+    // R4: only seed when both globalValue and defaultValue are defined and differ
+    if (inspect === undefined || inspect.globalValue === undefined || inspect.defaultValue === undefined) continue;
+    if (!isEqual(inspect.globalValue, inspect.defaultValue)) {
```
Comment (nick-y-snyk) [inline src/snyk/common/languageServer/explicitLsKeyTracking.ts:62]: The `defaultValue === undefined` guard silently skips settings that have no `default:` in `package.json` — specifically `snyk.advanced.organization`, `snyk.advanced.customEndpoint`, `snyk.advanced.cliPath`, and `snyk.advanced.additionalParameters`. If a user has any of these set, seeding is skipped and the first `workspace/didChangeConfiguration` sends `changed:false` — the same bug this PR fixes for other settings. T_F2 even asserts this behaviour as correct, but the PR description does not call it out as a known limitation. Either add a schema `default:` for these settings in `package.json`, or document this gap explicitly in the doc comment above.

Comment (bastiandoetsch) [reply src/snyk/common/languageServer/explicitLsKeyTracking.ts:62]: Fixed properly in 3e0ed5b8 rather than documenting it as a limitation. Removed the over-eager `defaultValue === undefined` clause from the guard: when a setting has no package.json default, `inspectConfiguration().defaultValue` is undefined and a defined `globalValue` is itself the deviation, so `isEqual(globalValue, undefined)` is false and the key is now seeded. `globalValue === undefined` is still skipped by the first guard, and settings that do have a default are unaffected.

Flipped T_F2 to assert these settings ARE seeded and added an integration test (T10) confirming `changed:true` propagates through `LanguageServerSettings.fromConfiguration`. Verified all four (organization, customEndpoint, cliPath, additionalParameters) are normal `SETTINGS_REGISTRY` entries (have `vscodeKey`, not `alwaysChanged`) that flow through the global changed-flag map, so seeding is the right fix here.

— AI review

**Comment by bastiandoetsch** [comment]
Good catch — agreed this is out of scope for this PR (as you note). The `package.json` vs `configuration.ts` defaults drift — particularly `issueViewOptions.ignoredIssues` — is a real risk worth fixing properly. I'll track it as a separate follow-up so `SETTINGS_REGISTRY.resolve()` fallbacks and the `DEFAULT_*` constants get reconciled into a single source of truth.

— AI review

**Comment by bastiandoetsch** [comment]
Filed as IDE-2135 (linked to IDE-2067) to reconcile the package.json ↔ configuration.ts defaults drift, with the `issueViewOptions.ignoredIssues` mismatch called out as the concrete example.

— AI review

**Review by nick-y-snyk — APPROVED** [review]
